### DISCUSSION TITLE
NeighborLoader/LinkNeighborLoader on Hetero Data

### DISCUSSION THREAD

**Discussion by colivarese** [pyg-team/pytorch_geometric#5269] [Q&A]
Hi, I'm currently trying to work with Heterogeneous Data, the data consists of two node types and ~900 edge types, I'm not sure how to make use of the mini-batch samplers with HeteroData, using the LinkNeighborLoader I have to input the edge_label_index, but I understand it is only possible to pass one edge_type, the same happens with edge_label. When I pass only one edge type it works,

`train_loader = LinkNeighborLoader(train_data,
                                num_neighbors = {key: [-1]*2 for key in train_data.edge_types }
                                ,shuffle = True,
                                  edge_label_index = train_data.edge_types[0],
                                  edge_label=None,
                                  batch_size = 128)`

The documentation indicates that I can pass a Tensor, EdgeType, or a Tuple with edge types and their indices, but wouldn't it only consider the labels for that particular edge type?

> edge_label_index (Tensor or EdgeType or Tuple[EdgeType, Tensor]) – The edge indices for which neighbors are sampled to create mini-batches. If set to [None](https://docs.python.org/3/library/constants.html#None), all edges will be considered. In heterogeneous graphs, needs to be passed as a tuple that holds the edge type and corresponding edge indices. (default: [None](https://docs.python.org/3/library/constants.html#None))

Or do I need to use the NeighborLoader? But I can only pass a node type into input_nodes.

**Comment by rusty1s** [comment]
This is still TBD, sorry. Currently, only single node types/edge types can be used as seed nodes. We have an open PR out to fix this, see https://github.com/pyg-team/pytorch_geometric/pull/5013.

**Reply by colivarese** [reply]
Thanks for the answer, It is good to know. I'll find a way to work it around, have a great day!

**Comment by MDrance** [comment]
Hello,
I have a question closely related to @colivarese one, I plan on doing link prediction on an hetero graph and something about `NeighborLoader` and `RandomLinkSplit` confuses me:
- `NeigborLoader` doesn't allow us, at the moment, to use multiple nodes types as seed nodes;
- `RandomLinkSplit` allow me to create train/val/test masks, and from the doc I can read: _edge_types (Tuple[EdgeType] or List[EdgeType], optional) – The edge types used for performing edge-level splitting in case of operating on [HeteroData](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.HeteroData) objects. (default: [None](https://docs.python.org/3/library/constants.html#None))_. So I can choose multiple types of edges to create the masks.

If I understood correctly, I could use this to predict links of different types as long as these links start from a node of type specified in `NeigborLoader` as my seeding. In this case, what would the masks look like ? One for each split considering all edges types or one mask per edge type ?

As an example, would this be correct ? 

```
train_data, val_data, test_data = T.RandomLinkSplit(
    num_val=0.1,
    num_test=0.1,
    neg_sampling_ratio=0.0,
    edge_types=[('user', 'rates', 'movie'), ('user', 'like', 'movie_genre')], #Two edges types
)(data)

loader = HGTLoader(
    hetero_data,
    num_samples={key: [512] * 4 for key in hetero_data.node_types},
    batch_size=128,
    input_nodes=('user', hetero_data['user'].train_mask), #Only one node type as seeding
)
```
In the case I would like to predict both links `('user', 'rates', 'movie')` and `('user', 'like', 'movie_genre')` from the only node type allowed in the `HGTLoader`, here `user` ?

My final goal is to train a model to predict multiple links of multiples types between multiple types of nodes, maybe there exists an other solution ?

Best,
Martin

**Reply by rusty1s** [reply]
I don‘t think this is correct. `RandomLinkSplit` does not create node-level masks - instead it splits edges into message passing edges and edges used for supervision (for each split). As such, you can use `RandomLinkSplit` to split your dataset for a link-level dataset, and use `LinkNeighborLoader` to create mini-batches to predict edge-level properties. Currently, we can only handle a single edge type as supervision though (as you correctly pointed out).

**Reply by MDrance** [reply]
Hello Matthias and thanks for your answer. 
I was wondering, is there some place I could find a community to chat with, a slack or maybe a discord of people working with GNN/Pytorch_geometric ?

Best,
Martin

**Reply by rusty1s** [reply]
Feel free to join our community Slack via https://data.pyg.org/slack.html

**Reply by wangjingyamap** [reply]
@rusty1s The slack link has expired.

**Reply by rusty1s** [reply]
Should be fixed.
